### DISCUSSION TITLE
400 instead of 404 for non-existing endpoint?

### DISCUSSION THREAD

**Discussion by devtud** [fastapi/fastapi#8147] [Questions] [answered]
**Description**
Let's say I have this endpoint: 
`GET /api/v1/users/{user_id}`

If I hit `GET /api/v1/users/6` and the user with id 6 does not exist, the response will be a 404. And that's correct.

However, if I misspell the endpoint and instead I hit `GET /api/v1/usersd/1` I still get a 404, even if `/api/v1/usersd/` is not declared anywhere.

As I find this misleading, I would like to return a 400 Bad Request for any request to a path that does not exist.

First of all, does this make sense? And how is this achievable?

**Comment by euri10** [comment]
I think you get the 404 directly from starlette if I'm correct:

https://www.starlette.io/routing/

> The router will respond with "404 Not found" or "405 Method not allowed" responses for requests which do not match.

**Comment by devtud** [comment]
@euri10 thank you!

I sort of found a solution and seems working. Just declare a generic endpoint in which I raise a 400, just after including the router with the real api endpoints.

And to make it "invisible", I don't include it in docs,

```
@app.get('/{path:path}', include_in_schema=False)
async def raise_404(path):
    raise HTTPException(status_code=400, detail='Bad request')
```

I'm also interested in better alternatives.

**Comment by tiangolo** [comment, marked as answer]
Hmm, I think this is not a very common use case, as 404 is used for URLs not found in the server, so 404 in both places would be "correct" as well.

An alternative I can think of is using a custom hint for not found users, maybe an extra code, or message.

E.g. having `404` but a message of `{"path_ok": True, "message": "User not found"}`.

But for your specific needs, your workaround might be the best solution.

---

As I understand you were able to solve your problem, I'll close this issue now. But feel free to add more comments or create new issues.

Also, you might ask in the Gitter chat: https://gitter.im/tiangolo/fastapi.
